### PR TITLE
RATIS-1761. If LeaderStateImpl is not running, it should not restart a LogAppender.

### PR DESCRIPTION
**Pull request by szetszwo**
See https://issues.apache.org/jira/browse/RATIS-1761

### CODE REVIEW

File: ratis-server/src/main/java/org/apache/ratis/server/impl/LeaderStateImpl.java [hunk 1]
```diff
@@ -393,7 +394,9 @@ boolean isReady() {
   }
 
   void stop() {
-    this.running = false;
+    if (!isStopped.compareAndSet(false, true)) {
+      return;
```
Comment (codings-dan) [inline ratis-server/src/main/java/org/apache/ratis/server/impl/LeaderStateImpl.java:398]: add log `LOG.warn("Failed to stop: can't set isStopped state");`

File: ratis-server/src/main/java/org/apache/ratis/server/impl/LeaderStateImpl.java [hunk 2]
```diff
@@ -595,8 +599,20 @@ void stopAndRemoveSenders(Predicate<LogAppender> predicate) {
...
+      return false;
+    }
+    final LeaderStateImpl current = server.getRole().getLeaderState().orElse(null);
+    return this == current;
+  }
+
   @Override
   public void restart(LogAppender sender) {
+    if (!isRunning()) {
+      return;
```
Comment (codings-dan) [inline ratis-server/src/main/java/org/apache/ratis/server/impl/LeaderStateImpl.java:613]: add log : `LOG.warn("Failed to restart: server is not running");`

File: ratis-server/src/main/java/org/apache/ratis/server/leader/LogAppenderBase.java [hunk 3]
```diff
@@ -125,6 +125,13 @@ public void stop() {
     daemon.tryToClose();
   }
 
+  void restart() {
+    if (!server.getInfo().isAlive()) {
+      return;
```
Comment (codings-dan) [inline ratis-server/src/main/java/org/apache/ratis/server/leader/LogAppenderBase.java:130]: add log: `LOG.warn("Failed to restart: server is not alive");`

**Review by codings-dan — COMMENTED** [review]
@szetszwo Thanks for working on this! Overall LGTM, please add some log for troubleshooting.

**Comment by szetszwo** [comment]
@codings-dan , thanks for reviewing this!  Just added some log messages.

**Review by codings-dan — APPROVED** [review]
the change looks good, +1

**Pull request merged**